scripts/clean_osm_data.py

```diff
@@ -0,0 +1,1832 @@
+    information on the substations shape and electrical parameters. Ways and
+    relations are subsequently concatenated to form a single DataFrame
+    containing unique bus ids.
+
+    Args:
+        path_substations (dict): A dictionary containing input paths for
+        substations.
+
+    Returns:
+        pd.DataFrame: A DataFrame containing the imported substations data.
+    """
+    cols_substations_way = [
+        "id",
+        "geometry",
+        "country",
+        "power",
+        "substation",
+        "voltage",
+        "frequency",
+    ]
+    cols_substations_relation = [
+        "id",
+        "country",
+        "power",
+        "substation",
+        "voltage",
+        "frequency",
+    ]
+    df_substations_way = pd.DataFrame(columns=cols_substations_way)
+    df_substations_relation = pd.DataFrame(columns=cols_substations_relation)
+
+    logger.info("Importing substations")
+    for key in path_substations:
+        logger.info(f"Processing {key}...")
+        for idx, ip in enumerate(path_substations[key]):
+            if (
+                os.path.exists(ip) and os.path.getsize(ip) > 400
+            ):  # unpopulated OSM json is about 51 bytes
+                country = os.path.basename(os.path.dirname(path_substations[key][idx]))
+                logger.info(
+                    f" - Importing {key} {str(idx+1).zfill(2)}/{str(len(path_substations[key])).zfill(2)}: {ip}"
+                )
+                with open(ip, "r") as f:
+                    data = json.load(f)
+
+                df = pd.DataFrame(data["elements"])
+                df["id"] = df["id"].astype(str)
+                # new string that adds "way/" to id
+                df["id"] = df["id"].apply(
+                    lambda x: (
+                        f"way/{x}" if key == "substations_way" else f"relation/{x}"
+                    )
+                )
+                df["country"] = country
+
+                col_tags = ["power", "substation", "voltage", "frequency"]
+
+                tags = pd.json_normalize(df["tags"]).map(
+                    lambda x: str(x) if pd.notnull(x) else x
+                )
+
+                for ct in col_tags:
+                    if ct not in tags.columns:
+                        tags[ct] = pd.NA
+
+                tags = tags.loc[:, col_tags]
+
+                df = pd.concat([df, tags], axis="columns")
+
+                if key == "substations_way":
+                    df.drop(columns=["type", "tags", "bounds", "nodes"], inplace=True)
+                    df_substations_way = pd.concat(
+                        [df_substations_way, df], axis="rows"
+                    )
+                elif key == "substations_relation":
+                    df.drop(columns=["type", "tags", "bounds"], inplace=True)
+                    df_substations_relation = pd.concat(
+                        [df_substations_relation, df], axis="rows"
+                    )
```

Comment:
For performance it could make sense to defer the concatenation until all dataframes to concatenate are collected in a list. But as the performance was very good, no action required.